### PR TITLE
Added VideoView Fixes #639

### LINKED ISSUE #639: Remove Implicit Intent From VideoWidget
The video widget calls an implicit intent when the video can be easily played in the app itself using a video view.
Moreover, this is inconsistent with the image view widget in which the image is loaded without any implicit intents.

### PR DESCRIPTION
**Pull request by rajatthakur**
Hey @lognaturel,
Please review the changes that I have made and let me know if anything else needs to be done. 

**Final Output**

I have ensured that if the VideoView fails to load the video file, the video is loaded through an implicit intent.
Also,I have tested this on an actual device as well as on an emulator.
**Thank You**

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java [hunk 1]
```diff
@@ -193,6 +200,18 @@ public void onClick(View v) {
         mPlayButton.setPadding(20, 20, 20, 20);
         mPlayButton.setLayoutParams(params);
 
+        //setup video view and media controller
+        mVideoView = new VideoView(getContext());
+        mVideoView.setId(QuestionWidget.newUniqueId());
+        mVideoView.setPadding(20,20,20,20);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java]: Please use spaces after commas to let the code breathe!

File: collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java [hunk 2]
```diff
@@ -193,6 +200,18 @@ public void onClick(View v) {
...
+        mVideoView.setVisibility(View.VISIBLE);
+        DisplayMetrics metrics = context.getResources().getDisplayMetrics();
+        mVideoView.setLayoutParams(new FrameLayout.LayoutParams
+                (ViewGroup.LayoutParams.MATCH_PARENT, (metrics.heightPixels*4)/10));
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java]: Spaces after commas and between operands in expressions.

**Review by lognaturel — CHANGES_REQUESTED** [review]
Looks good overall and just needs a couple of stylistic changes.

I think the user interface could be refined a little bit but perhaps we can file this as one or more separate issues:
- don't show a black box if not video has been selected
- when the field is cleared (long press), remove the video
- if possible, show a frame from the video rather than a black box (like when the video has been played once)

**Comment by rajatthakur** [comment]
Hey @lognaturel ,
I have pushed in a commit as you had requested (with all the three issues fixed).
Thank You :smile:

**Comment by lognaturel** [comment]
Thanks, @rajatthakur! I'm still trying to figure out whether this is truly a better user experience. @nribeka @yanokwa you should take a look. 

@rajatthakur is there a way to give an option to make the video full screen?

**Comment by rajatthakur** [comment]
Hey @lognaturel,
I have added an option to make the video fullscreen as you had asked.

**Final Result**

**Comment by yanokwa** [comment]
I like the fact that it takes up the width of the screen. Great job there!

I don't like how full screen is done with a button in the Collect UI. Is there an easy way to do this inline with the player? Something like the far right button in the YouTube button?
youtube

**Comment by rajatthakur** [comment]
Hi @yanokwa,
I have added the fullscreen functionality in the media controller itself as you had asked.

 

Thank You :smile:

**Comment by rajatthakur** [comment]
Hi @lognaturel ,
Please review this and let me know if any changes need to be made.

Thank You :smile:

**Comment by lognaturel** [comment]
The "full screen" button is a neat addition, @rajatthakur. What I'm worried about is that it still doesn't actually let the video be full screen. That may make it hard for a data collector to review a video that needs to have certain details. @yanokwa, @nribeka, do you think this is a user experience improvement over launching an external player?

**Comment by yanokwa** [comment]
Hi @rajatthakur! Thanks for sending in this PR and I'm sorry I haven't been able to give you timely feedback. I just tried this out and I think it's a great idea, but it needs more work before it can be accepted into the project. Here's what I'd expect to see for this PR:

1. If there is a video stored, you should replace the Play Video button with an embedded player. Currently, we always see Play Video button.

1. The size of the embedded video should change when the screen is rotated to fill the screen. Currently, the video is the correct size in portrait mode, but it doesn't fill the screen when in landscape mode (although the scrubber does).

1. There should be an option to play the video at full-screen. Right now, the full-screen option seems to replace the Record/Choose/Play buttons. I think what would be ideal is for that button launch the stock player so you get all the functionality of the current video implementation (share, info, etc).

Again, I think you're on the right track and with a little bit more work, this feature can go out to our users. Thanks!

**Comment by rajatthakur** [comment]
Hey @yanokwa 
First of all, I am really sorry for replying so late. I had a couple of exams which kept me really occupied.
I have pushed in a commit with a few of the changes that you had suggested. I have removed the PlayButton as you had asked. As soon as the user selects a video or selects the video widget again after having recorded the response, the video view shows up and the user can play, pause, seek etc. I could have kept the play button when the response is selected for the first time, but that was making the code unnecessarily complicated (because the view was getting refreshed several times). 
The full-screen button has been modified to call an intent to an external media player.
About the landscape issue, I don't think the video width can get any bigger given the limited amount of screen available for video view in landscape mode. I can probably stretch the video to fill the entire width but that according to me wouldn't look nice. So, I have not changed anything with respect to this point in this commit.
Please let me know if any other changes need to be made.
Thank You :smile:

**Comment by yanokwa** [comment]
@rajatthakur I haven't had a chance to look at this, but can you please make sure that your code passes lint and style checks? If you are comfortable with the command line, run `./gradlew checkstyle` and `./gradlew lintDebug` (or the checkstyle and lintDebug tasks in Android Studio).

**Comment by yanokwa** [comment]
The more I play with this PR, then more I think it's an inconsistent user experience. For example, if a user records in portrait view, the scrubber takes up the whole width of the screen and that visually looks strange. Also strange is that the video doesn't take up all the available vertical space. And if a user records in landscape and views in landscape, then their scrubber obscures the video they need to watch. Also, choosing a saved video results in a crash.

All this iteration takes time, and without a clear proposal in the issue about the high level problem we are trying to solve, I think we are burning a lot of time. So my apologies, @rajatthakur, I know you've been working hard on this, but I think we should close this PR for now. And extra apologies, because I should have thought through and made the decision earlier.

I still think the Play Video button in the current tools is not ideal, but before we can review another PR, there should be a clear proposal in the issue about how an inline player will work for videos captured and viewed in landscape and portrait and why that behavior will be better that current practice. And we should have rough sketches of all that before any code is written.

**Pull request closed without merging**